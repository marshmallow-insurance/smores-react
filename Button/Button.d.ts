import React, { ButtonHTMLAttributes, FC, ReactNode } from 'react';
import { MarginProps } from '../utils/space';
type Props = {
    children: ReactNode;
    id?: string;
    className?: string;
    disabled?: boolean;
    handleClick?: (e: React.FormEvent<HTMLButtonElement>) => void;
    loading?: boolean;
    primary?: boolean;
    secondary?: boolean;
    fallbackStyle?: boolean;
    textBtn?: boolean;
    smallButton?: boolean;
    icon?: string;
    trailingIcon?: boolean;
    forcedWidth?: string;
    form?: string;
};
export type ButtonProps = Props & MarginProps & Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'color'>;
export declare const Button: FC<ButtonProps>;
export {};
