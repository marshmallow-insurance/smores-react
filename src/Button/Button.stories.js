import React from 'react';
import {storiesOf} from '@storybook/react';

import {text} from '@storybook/addon-knobs';

import {Button} from './Button';

storiesOf('Button', module)
  .addWithJSX('default', () => <Button color="blue">Me Button</Button>)
  .addWithJSX('coloured', () => (
    <Button color={text('Color', 'green')}>Coloured Button</Button>
  ))
  .addWithJSX('inverted', () => <Button inverted>Me Button</Button>)
  .addWithJSX('outlined', () => <Button outlined>Outlined</Button>)
  .addWithJSX('disabled', () => (
    <Button color={text('Color', 'blue')} disabled>
      Me Button
    </Button>
  ))
  .addWithJSX('block', () => <Button block>Me Button</Button>)
  .addWithJSX('clickable', () => (
    <Button handleSubmit={() => console.log('clicked')}>Click me</Button>
  ));
