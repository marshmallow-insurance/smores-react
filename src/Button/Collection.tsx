import React, { FC, ReactNode } from 'react'
import styled from 'styled-components'

import { Text } from '../Text'
import { Box } from '../Box'
import { Button, ButtonProps } from './Button'

export const buttonList: Array<ButtonProps> = [
  { primary: true, children: 'Save' },
  { secondary: true, children: 'Edit' },
  { fallback: true, children: 'Cancel' },
  { textBtn: true, children: 'View' },
  { smallButton: true, children: 'Small' },
]

export const CollectionPage: FC = () => {
  return (
    <Wrapper>
      <Section label="Default">
        <Row label="Default">
          {buttonList.map((props, index) => (
            <Button key={index} {...props} />
          ))}
        </Row>
        <Row label="Loading">
          {buttonList.map((props, index) => (
            <Button key={index} {...props} loading />
          ))}
        </Row>
        <Row label="Disabled">
          {buttonList.map((props, index) => (
            <Button key={index} {...props} disabled />
          ))}
        </Row>
      </Section>
      <Section label="With icon">
        <Row label="Default">
          {buttonList.map((props, index) => (
            <Button key={index} {...props} icon="info" />
          ))}
        </Row>
        <Row label="Loading">
          {buttonList.map((props, index) => (
            <Button key={index} {...props} loading icon="info" />
          ))}
        </Row>
        <Row label="Disabled">
          {buttonList.map((props, index) => (
            <Button key={index} {...props} disabled icon="info" />
          ))}
        </Row>
      </Section>
    </Wrapper>
  )
}

const Wrapper = styled(Box)`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 40px;
`

const Section: FC<{ label: string; children?: ReactNode }> = ({
  label,
  children,
}) => {
  return (
    <Box>
      <Box mb="16px">
        <Text tag="span" typo="header-small">
          {label}
        </Text>
      </Box>
      <Column>{children}</Column>
    </Box>
  )
}

const Column = styled(Box)`
  display: flex;
  flex-direction: column;
  gap: 16px;
`

const Row: FC<{ label: string; children?: ReactNode }> = ({
  label,
  children,
}) => {
  return (
    <RowWrapper>
      <Box width="80px">
        <Text tag="span" typo="label-large">
          {label}
        </Text>
      </Box>
      {children}
    </RowWrapper>
  )
}

const RowWrapper = styled(Box)`
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 16px;
`
