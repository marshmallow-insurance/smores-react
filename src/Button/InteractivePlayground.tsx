import React, { FC, useState } from 'react'

import { Button } from './Button'

function sleep(delay: number) {
  return new Promise(function (resolve) {
    setTimeout(resolve, delay)
  })
}

export const InteractivePlayground: FC = () => {
  const [isLoading, setIsLoading] = useState(false)

  const handleClick = async () => {
    setIsLoading(true)
    await sleep(1000)
    setIsLoading(false)
  }

  return (
    <Button primary icon="info" handleClick={handleClick} loading={isLoading}>
      Get started
    </Button>
  )
}
