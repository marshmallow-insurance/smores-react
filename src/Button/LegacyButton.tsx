import React, { FC, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import { theme } from '../theme'

interface IButton {
  /** button color  */
  color: string
  /** unique id */
  id: string
  /** take full 100% width  */
  block: boolean
  /** invert bg and text colors */
  inverted: boolean
  /** disabled state */
  disabled: boolean
  /** outline styling */
  outlined: boolean
  /** onClick event handler */
  onClick: (e: React.FormEvent<HTMLButtonElement>) => void
}

type Props = {
  children: ReactNode
  id: string
  className?: string
  color?: string
  block?: boolean
  inverted?: boolean
  disabled?: boolean
  outlined?: boolean
  handleClick: (e: React.FormEvent<HTMLButtonElement>) => void
}

export const LegacyButton: FC<Props> = ({
  children,
  id,
  className = '',
  color = 'blue',
  block = false,
  inverted = false,
  disabled = false,
  outlined = false,
  handleClick,
}) => (
  <Container
    id={id}
    className={className}
    color={color}
    block={block}
    inverted={inverted}
    disabled={disabled}
    outlined={outlined}
    onClick={handleClick}
  >
    {children}
  </Container>
)

const Container = styled.button<IButton>(
  ({ block, color, inverted, outlined }) => css`
    position: relative;
    display: inline-block;
    box-sizing: border-box;
    border: none;
    border-radius: 8px;
    font-size: 16px;
    padding: 18px 16px 14px;
    outline: none;
    cursor: pointer;
    width: ${block ? '100%' : 'auto'};
    background-color: ${theme.colors[`${color}5`]};
    color: ${theme.colors.white};
    font-weight: ${theme.font.weight.medium};

    &:hover:not([disabled]) {
      background-color: ${theme.colors[`${color}6`]};
    }
    &:active:not([disabled]) {
      background-color: ${theme.colors[`${color}7`]};
    }
    &:disabled {
      opacity: 0.5;
      cursor: not-allowed;
    }

    ${(inverted || outlined) &&
    css`
      background-color: transparent;
      border: 1px solid ${outlined ? theme.colors.grey4 : 'transparent'};
      color: ${outlined ? theme.colors.blue7 : theme.colors[`${color}5`]};

      &:hover:not([disabled]) {
        background-color: ${theme.colors.bg2};
      }
      &:active:not([disabled]) {
        background-color: ${theme.colors.bg3};
      }
    `};

    @media (min-width: 768px) {
      padding: 16px 16px 15px;
      font-size: 16px;
    }
  `,
)
