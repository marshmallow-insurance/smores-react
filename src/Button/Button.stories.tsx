import React from 'react'

import { Button, ButtonProps } from './Button'
import { InteractivePlayground } from './InteractivePlayground'
import { CollectionPage } from './Collection'

export default {
  title: 'Button',
  component: Button,
  argTypes: { handleClick: { action: 'clicked' } },
}

const Template = (props: ButtonProps) => <Button {...props}>Get started</Button>

const InteractivePlaygroundTemplate = (props: ButtonProps) => (
  <InteractivePlayground {...props} />
)

export const Playground = Template.bind({})

Playground.args = {
  primary: true,
  secondary: false,
  fallback: false,
  disabled: false,
  loading: false,
  icon: '',
}

export const Interactive = InteractivePlaygroundTemplate.bind({})
export const Collection = CollectionPage.bind({})

export const ForcedWidth = Template.bind({})

ForcedWidth.args = {
  primary: true,
  forcedWidth: '300px',
}
