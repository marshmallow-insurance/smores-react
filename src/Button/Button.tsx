import React, { FC, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import { theme } from '../theme'
import { LegacyButton } from './LegacyButton'
import { Loader } from '../Loader'
import { Icon as IconComponent } from '../Icon'

interface IButton {
  /** disabled state */
  disabled: boolean
  /** loading state */
  loading: boolean
  /** primary button styling */
  primary: boolean
  /** secondary button styling */
  secondary: boolean
  /** tertiary button styling */
  tertiary: boolean
  /** icon state */
  icon: string
  /** forcedWidth styling */
  forcedWidth: string
}

type Props = {
  children: ReactNode
  id: string
  className?: string
  color?: string
  block?: boolean
  inverted?: boolean
  disabled?: boolean
  outlined?: boolean
  handleClick: (e: React.FormEvent<HTMLButtonElement>) => void
  loading?: boolean
  primary?: boolean
  secondary?: boolean
  tertiary?: boolean
  icon?: string
  forcedWidth?: string
}

export const Button: FC<Props> = ({
  children,
  id,
  className = '',
  color = 'blue',
  block = false,
  inverted = false,
  disabled = false,
  outlined = false,
  handleClick,
  loading = false,
  primary = false,
  secondary = false,
  tertiary = false,
  icon = '',
  forcedWidth = '',
}) => {
  return (
    <div>
      {primary || secondary || tertiary ? (
        <Container
          id={id}
          disabled={disabled}
          onClick={handleClick}
          loading={loading}
          primary={primary}
          secondary={secondary}
          tertiary={tertiary}
          icon={icon}
          forcedWidth={forcedWidth}
        >
          {icon && (
            <IconContainer
              render={icon}
              size={24}
              color={primary ? 'white' : `${theme.colors.blue7}`}
            />
          )}
          {loading ? (
            <Loader
              color={primary ? 'white' : `${theme.colors.blue7}`}
              height="16"
            />
          ) : (
            children
          )}
        </Container>
      ) : (
        <LegacyButton
          id={id}
          className={className}
          color={color}
          block={block}
          inverted={inverted}
          disabled={disabled}
          outlined={outlined}
          handleClick={handleClick}
        >
          {children}
        </LegacyButton>
      )}
    </div>
  )
}

const Container = styled.button<IButton>(
  ({
    disabled,
    loading,
    primary,
    secondary,
    tertiary,
    icon,
    forcedWidth,
  }) => css`
    background-color: ${theme.colors.pink5};
    box-shadow: none;
    color: ${theme.colors.blue7};
    padding: 16px 20px;
    border: none;
    outline: none;
    border-radius: 8px;
    align-items: center;
    display: flex;
    justify-content: ${icon ? 'space-evenly' : 'center'};
    font-weight: ${theme.font.weight.medium};
    cursor: ${disabled || loading ? 'not-allowed' : 'pointer'};
    line-height: 100%;
    font-size: 16px;
    opacity: ${disabled ? '0.5' : '1'};
    width: ${forcedWidth ? forcedWidth : 'auto'};

    ${primary &&
      css`
        color: ${theme.colors.white};

        &:hover {
          background-color: ${!(disabled || loading) && theme.colors.pink6};
        }
        &:active {
          background-color: ${!(disabled || loading) && theme.colors.pink7};
        }
      `}
    ${secondary &&
        css`
        background-color: ${theme.colors.white};
        border: 2px solid ${theme.colors.blue7};

        &:hover {
          background-color: ${!(disabled || loading) && theme.colors.bg2};
          border: ${!(disabled || loading) && `2px solid ${theme.colors.blue6}`};
        }
        &:active {
          background-color: ${theme.colors.bg2};
          border: 2px solid ${theme.colors.blue7};
        }
      `}
  ${tertiary &&
    css`
      background-color: ${theme.colors.bg2};
      
      &:hover {
        background-color: ${!(disabled || loading) && theme.colors.grey2};
      }
      &:active {
        background-color: ${theme.colors.grey3};
      }
    `}
  `,
)

const IconContainer = styled(IconComponent)`
  padding-right: 10px;
`
