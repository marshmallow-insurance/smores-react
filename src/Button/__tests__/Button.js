import React from 'react'
import { render } from '@testing-library/react'
import 'jest-styled-components'

import { Button } from '../Button'

test('renders', () => {
  const { container } = render(<Button color="green" />)
  expect(container.firstChild).toMatchSnapshot()
})
