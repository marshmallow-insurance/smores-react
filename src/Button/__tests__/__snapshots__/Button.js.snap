// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c0 {
  position: relative;
  background-color: #FF88C8;
  box-shadow: none;
  color: #292924;
  padding: 0 20px;
  outline: none;
  border-radius: 10000px;
  font-weight: 500;
  cursor: pointer;
  line-height: 100%;
  font-size: 16px;
  opacity: 1;
  width: auto;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 1;
}

.c2 {
  padding: 16px 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<button
  class="c0"
  color="green"
>
  <div
    class="c1"
  >
    <div
      class="c2 childrenContainer"
    />
  </div>
</button>
`;
